### PR TITLE
Add new APIs to ExoMediaDrm

### PR DESCRIPTION
**Pull request by shanujshekhar**
Changes
---
- Added `removeOfflineLicense(byte[])` and `getOfflineLicenseKeySetIds` and consumed them in their implementations

Background
---
- These APIs will help in addressing an increasing amount of `java.lang.IllegalArgumentException: Failed to restore keys: BAD_VALUE` which is our top playback error in our app
	- Based on our discussion with Widevine team and [this exoplayer issue](https://github.com/google/ExoPlayer/issues/11202#issuecomment-1708792594)
		- TL;DR: The failure occurs on startup if the user has 200+ offline licenses, we would like to add the functionality to remove offline licenses

**Note: Why we want these APIs in ExoMediaDrm and not in OfflineLicenseHelper**

- As per the issue above, we would like to access these 2 public APIs in MediaDrm that don’t exist in `OfflineLicenseHelper` or `ExoMediaDrm`
	- APIs interested in:
		- [MediaDrm#removeOfflineLicense()](https://developer.android.com/reference/android/media/MediaDrm#removeOfflineLicense(byte%5B%5D)): To remove offline license
		- [MediaDrm#getOfflineLicenseKeySetIds()](https://developer.android.com/reference/android/media/MediaDrm#getOfflineLicenseKeySetIds()): To see number of offline licenses on startup

	- We use `OfflineLicenseHelper` to download license for L1 and we don't interact with `ExoMediaDrm` directly. But for the alternate Widevine integration, we directly depend on `ExoMediaDrm` APIs to override and call CDM Native APIs.
	- We would like to have the functionality of removing offline licenses for both integration which would need access to above APIs in `ExoMediaDrm`.

Links
---
- https://github.com/androidx/media/issues/659

### CODE REVIEW

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java [hunk 1]
```diff
@@ -583,4 +583,14 @@ byte[] provideKeyResponse(byte[] scope, byte[] response)
    */
   @C.CryptoType
   int getCryptoType();
+
+  /**
+   * See {@link MediaDrm#removeOfflineLicense(byte[])}
```
Comment (icbaker) [inline libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java:588]: It's OK to `@see` or `See @link ...` to the `MediaDrm` docs in a second line, but let's make the first line of this javadoc stand alone.

I also think it's worth flagging the caveat that this method is generally not needed here too.

So something like: 
> Removes an offline license.
>
> This method is generally not needed, and  should only be used if the preferred approach of passing {@link KEY_TYPE_RELEASE} to {@link #getKeyRequest} is not possible.
>
> See {@link MediaDrm.removeOfflineLicense} for more details.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java [hunk 2]
```diff
@@ -380,6 +382,21 @@ public FrameworkCryptoConfig createCryptoConfig(byte[] sessionId) throws MediaCr
     return C.CRYPTO_TYPE_FRAMEWORK;
   }
 
+  @Override
+  public void removeOfflineLicense(byte[] keySetId) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
```
Comment (icbaker) [inline libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java:387]: I don't think it's really acceptable to make this method a no-op on pre-Q devices without making that much clearer in the interface.

See for example the docs on `setOnKeyStatusChangeListener` and the implementation that throws an exception on unsupported API levels.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java [hunk 3]
```diff
@@ -156,4 +156,14 @@ public CryptoConfig createCryptoConfig(byte[] sessionId) {
...
   }
+
+  @Override
+  public void removeOfflineLicense(byte[] keySetId) {
+    // Do nothing
+  }
+
+  @Override
+  public List<byte[]> getOfflineLicenseKeySetIds() {
+    return null;
```
Comment (icbaker) [inline libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java:167]: This method is not annotated `@Nullable`, so it's implicitly non-null. Probably best to return an empty list here.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java [hunk 4]
```diff
@@ -156,4 +156,14 @@ public CryptoConfig createCryptoConfig(byte[] sessionId) {
   public @C.CryptoType int getCryptoType() {
     return C.CRYPTO_TYPE_UNSUPPORTED;
   }
+
+  @Override
+  public void removeOfflineLicense(byte[] keySetId) {
+    // Do nothing
```
Comment (icbaker) [inline libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java:162]: For consistency with methods above that take `keySetId` (like `getKeyRequest`) this should probably throw an exception

**Review by icbaker — CHANGES_REQUESTED** [review]

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java (hunk 1 above)
Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java]: Thanks for reviewing. Updated.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java (hunk 2 above)
Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java]: Thanks for reviewing. Updated.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java (hunk 3 above)
Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java]: Thanks for reviewing. Updated.

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java (hunk 4 above)
Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DummyExoMediaDrm.java]: Thanks for reviewing. Updated.

**Review by icbaker — APPROVED** [review]

**Comment by shanujshekhar** [comment]
Hey Ian, nothing urgent but I had a follow-up question I posted in [the issue](https://github.com/androidx/media/issues/659#issuecomment-1734032140), not sure if you had a chance to take a look but it would be really great if you could help in that regard. Thank you!

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java [hunk 5]
```diff
@@ -327,6 +329,24 @@ public void restoreKeys(byte[] sessionId, byte[] keySetId) {
...
+      throw new UnsupportedOperationException();
+    }
+    mediaDrm.removeOfflineLicense(keySetId);
+  }
+
+  @Override
+  @RequiresApi(Build.VERSION_CODES.Q)
+  public List<byte[]> getOfflineLicenseKeySetIds() {
+    if (Util.SDK_INT < Build.VERSION_CODES.Q) {
+      return ImmutableList.of();
```
Comment (shanujshekhar) [inline libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java:345]: Thinking if we should also throw `UnsupportedOperationException()` exception incase of unsupported API level to match the pattern

Comment (icbaker) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java:345]: Yes you're right, that is easier to reason about and more consistent - done.

Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/FrameworkMediaDrm.java:345]: Got it. Thank you!

File: libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java [hunk 6]
```diff
@@ -523,6 +523,32 @@ byte[] provideKeyResponse(byte[] scope, byte[] response)
...
   void restoreKeys(byte[] sessionId, byte[] keySetId);
 
+  /**
+   * Removes an offline license.
+   *
+   * <p>This method is generally not needed, and should only be used if the preferred approach of
+   * generating a license release request by passing {@link #KEY_TYPE_RELEASE} to {@link
+   * #getKeyRequest} is not possible.
+   *
+   * <p>This is an optional method, and some implementations may only support it on certain Android
```
[the remark this answers is not in the document]
Comment (icbaker) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java:533]: Good point, done on both the new methods.

[the remark this answers is not in the document]
Comment (shanujshekhar) [reply libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/ExoMediaDrm.java:533]: Thank you so much!

**Pull request merged**